### PR TITLE
CLIENT-SPECIFICATION, style-guide: add longform/shortform specifications

### LINKED ISSUE #13556: Proposal: Interpret {{[*|*]}} in the client
As discussed in https://github.com/tldr-pages/tldr/issues/13466, this is to make it user configurable to show either longform or shortform options. It would allow for more pasteability. This would also be backwards compatible as we already use that notation. It would not break any existing clients.

### PR DESCRIPTION
**Pull request by Managor**
See these for more information:
#13556
https://github.com/tldr-pages/tldr-python-client/pull/259

Closes #13556.

### CODE REVIEW

**Comment by Managor** [comment]
For now, I'm looking for feedback before opening for review.

**Review by acuteenvy — COMMENTED** [review]
The current syntax (i.e. `{{-s|--long}}`) doesn't work for programs that don't follow this convention for options (e.g. `nuclei` - example short|long option placeholder: `{{-l|-list}}`, and now there's no way for a client to distinguish this from an ordinary choice placeholder). If we standardize this, it's going to be hard to change it to support different option conventions in the future, hence I'd like to propose an alternative syntax.

#5092 was a proposal to use `{[ ... ]}` for option selection. That is not backwards compatible, because it uses a single curly brace, which makes it not a placeholder by the current specification.
What if we combine the two (i.e. use `{{[-s|--long]}}`)? Current clients will display `[-s|--long]` with placeholder styling, which still looks like a choice, and on top of that a short option doesn't need to start with a single hyphen, and a long option doesn't need to start with two hyphens. As long as the square brackets are included, an option can be whatever we want it to be, allowing even for subcommand short forms (e.g. `cargo {{[b|build]}}`).

What do you think about this?

**Comment by Managor** [comment]
After mulling it over, I've come to accept the idea. There needs to be flexibility to add arbitrary shortform/longform arguments.

Do you think it would be good to make the expected default behavior to only show the longform options. It would remove clutter for new users. I don't think having `-s|--longform` plastered everywhere is a very newbie friendly.

File: CLIENT-SPECIFICATION.md [hunk 1]
```diff
@@ -39,9 +39,13 @@ Option             | Required?   | Meaning
 `-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if caching is supported.
 `-l`, `--list`     | No          | Lists all the pages in the current platform to the standard output.
 `-L`, `--language` | No          | Specifies the preferred language for the page returned. Overrides other language detection mechanisms. See the [language section](#language) for more information.
+`-S`, `--shortform`| No          | If set, will filter {{[*|*]}} patterns in examples to show their shortform option (first element).
+`-V`, `--longform` | No          | If set, will filter {{[*|*]}} patterns in examples to show their longform option (second element).
 
 Clients MUST implement both the short and long versions of an option.
```
Comment (Managor) [inline CLIENT-SPECIFICATION.md:45]: Does this need clarification? The MUST appearing right after `--shortform` and `--longform` options might cause confusion.

Comment (sbrl) [reply CLIENT-SPECIFICATION.md:45]: Yes, it is important that clients support both the short and the long forms of all arguments.

It is a MUST to ensure consistency of CLI experience ref `tldr tldr` etc. I strongly advise that it is left as a MUST.

Any particular reason for `-V` here as the short form of `--longform`?

Comment (Managor) [reply CLIENT-SPECIFICATION.md:45]: > Yes, it is important that clients support both the short and the long forms of all arguments.

You misunderstand me. What will happen when someone reads the specs, reads about shortform/longform options that are optional according to the table and then immediately after reads they you "MUST implement both the short and long versions of an option"? I would assume confusion.
I was asking if the wording needs clarification so that it's apparent that it's talking about the shortform/longform pair and not `--shortform` and `--longform`

> Any particular reason for -V here as the short form of --longform?

`-l` and `-L` were taken. I went for the next best option `-V`  short for verbose.

File: CLIENT-SPECIFICATION.md [hunk 2]
```diff
@@ -39,9 +39,13 @@ Option             | Required?   | Meaning
...
 
+By default clients SHOULD display only the longform option when neither `--shortform` or `--longform` is set by the user. If both are provided, both patterns should be displayed separated by a `|`.
```
Comment (acuteenvy) [inline CLIENT-SPECIFICATION.md:47]: Clients as they are now will display `{{[-s|--long]}}` as `[-s|--long]`. According to this specification, when both options are requested, they should display `-s|--long`.

Is this intended?

```suggestion
By default clients SHOULD display only the longform option when neither `--shortform` or `--longform` is set by the user. If both are provided, both patterns should be displayed, enclosed in `[ ]`, separated by a `|` (e.g. `[-s|--long]`).
```

Comment (Managor) [reply CLIENT-SPECIFICATION.md:47]: Great point! Currently how I designed it was so that the program removes the brackets when both are shown. See here https://github.com/tldr-pages/tldr-python-client/pull/259#issuecomment-2619942873
As you can see it isn't exactly clear when the option placeholder ends.

Compare that to the way tealdeer displays its placeholders with an underscore
image

Would it be ok to use the MAY keyword and let the client decide whether they want to leave the brackets in for clarity?

Comment (acuteenvy) [reply CLIENT-SPECIFICATION.md:47]: In my opinion, it would be better to standardize a single way of handling this, but let's see what others think.

File: CLIENT-SPECIFICATION.md (hunk 1 above)
Comment (sbrl) [reply CLIENT-SPECIFICATION.md:45]: Hmmm, I see your reasoning.

I think the wording "of an option" us unchanged from the time I originally wrote the client spec, though perhaps it would be clearer to say "of each given argument" instead?

Comment (sbrl) [reply CLIENT-SPECIFICATION.md:45]: Also ref `-V`, I can't help but think that could cause confusion. What about `-e` for expanded form, or even completely different `-a` and `-A` for short and long forms of arguments respectively?

File: CLIENT-SPECIFICATION.md (hunk 2 above)
Comment (sbrl) [reply CLIENT-SPECIFICATION.md:47]: Sitting down to this properly, I can see the argument that having a single standardised way to handle this is desirable. However, different groups of users will likely want something different here.

For example, a novice may appreciate having the long form shown, but a expert they prefer having the short form shown.

As it is right now, with the options described there is no way to show *both* at once, which could be useful for memorising the short argument on commonly used commands.

To this end, I would suggest it would be most beneficial to show *both* options at once, and then allow selective restriction with cli args, which could easily be e.g. saved as preferences on a client-by-client basis or otherwise `alias tldr=`tldr --arg` etc etc etc.

Also, this bit about presentation doesn't seem to fit here and feels rather out of place.. Perhaps it would do better in the [Page structure](https://github.com/tldr-pages/tldr/pull/15253/files#diff-58189d1a0aed9ae655ef8f782b9f1756ab9e0722d7c2e04a8479acc59d063f80L125) section?

Comment (Managor) [reply CLIENT-SPECIFICATION.md:47]: > As it is right now, with the options described there is no way to show both at once, which could be useful for memorising the short argument on commonly used commands.
What do you mean? "If both are provided, both patterns should be displayed"
See here for examples https://github.com/tldr-pages/tldr-python-client/pull/259#issuecomment-2619942873

I have also implemented environment variable in the python client, but I didn't see any other envar in the spec.
Most I could find in on the python client page https://github.com/tldr-pages/tldr-python-client/pull/259#issuecomment-2619942873

Not sure how I'd split this into the option part and page structure part. Can you offer a suggestion?

File: CLIENT-SPECIFICATION.md (hunk 1 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:45]: `-o` for shortform and `-O` for longform. Short for "option"?

Comment (Managor) [reply CLIENT-SPECIFICATION.md]: I moved the long+short option requirement to the start. I assume that is much clearer with its intentions.

**Comment by Managor** [comment]
Is there any way to disable github actions for this pull request?. It's getting in the way of reading what has changed.

File: CLIENT-SPECIFICATION.md (hunk 2 above)
Comment (sbrl) [reply CLIENT-SPECIFICATION.md:45]: > "If both are provided, both patterns should be displayed"

Oops, didn't see that. But what I'm trying to say is that there's an argument for both defaulting to long options and defaulting to short options.

> Not sure how I'd split this into the option part and page structure part. Can you offer a suggestion?

Consider moving the discussion of the syntax (i.e. `{{[...]}}`) to the page structure heading?

Comment (Managor) [reply CLIENT-SPECIFICATION.md]: Done.

File: CLIENT-SPECIFICATION.md [hunk 3]
```diff
@@ -117,7 +115,7 @@ The structure inside these translation folders is identical to that of the main
...
... with the exception of the non-standard `{{`, `}}`, `{{[` and `]}}` placeholder syntax. `{{` and `}}` surrounds values in an example that indicates editable values. `{{[` and `]}}` indicate shortform and longform variants of options which are separated by a single `|`. Shortform on the left, longform on the right. Clients MAY highlight the placeholders and MUST remove the surrounding curly braces. Clients MUST remove the angle brackets from option placeholders when only short- or longform is shown and MAY remove them when both are shown depending whether it's clear where the the option placeholder starts and ends (for example when underlining placeholders). Clients MUST NOT treat them as the placeholder syntax if they are escaped using `\` (i.e. `\{\{` and `\}\}`) and MUST instead display literal braces, without backslashes. Placeholder escaping applies only when both braces are escaped (e.g. in `\{` or `\{{`, backslashes MUST be displayed). In cases when a command uses `{}` in its arguments (e.g. `stash@{0}`) **_the outer braces_** mark the placeholder - the braces inside MUST be displayed. Clients MUST NOT break if the page format is changed within the _CommonMark_ specification.
```
Comment (sbrl) [inline CLIENT-SPECIFICATION.md:118]: This is much better being moved down here!

This paragraph is getting pretty long though, so I wonder if we could break it up a bit? Perhaps into multiple smaller paragraphs, or perhaps bullet points?

Comment (Managor) [reply CLIENT-SPECIFICATION.md]: Done. That's much more readable. Anything more to add?

File: CLIENT-SPECIFICATION.md [hunk 4]
```diff
@@ -54,6 +56,17 @@ tldr --version
...
+
+The following environment variables MAY be supported by the client to alter its behavior:
+
+```
+TLDR_PLATFORM
+TLDR_LANGUAGE
+TLDR_SHORTFORM
+TLDR_LONGFORM
+```
+
```
Comment (Managor) [inline CLIENT-SPECIFICATION.md:69]: I added an environment variable section

File: CLIENT-SPECIFICATION.md [hunk 5]
```diff
@@ -39,8 +39,10 @@ Option             | Required?   | Meaning
 `-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if caching is supported.
 `-l`, `--list`     | No          | Lists all the pages in the current platform to the standard output.
 `-L`, `--language` | No          | Specifies the preferred language for the page returned. Overrides other language detection mechanisms. See the [language section](#language) for more information.
+`-o`, `--shortform`| No          | If set, will filter examples to show their shortform option when available
+`-O`, `--longform` | No          | If set, will filter examples to show their longform option when available
```
Comment (acuteenvy) [inline CLIENT-SPECIFICATION.md:43]: Before we merge this, I would like to ask everyone for their opinion on these short options, because changing them in all clients later on will be very hard to do. Which ones do you think are better?
@kbdharun @sebastiaanspeck @spageektti

- `-S` for short, `-V` for long options (original proposal)
- `-o` for short, `-O` for long options (current)
discussed in https://github.com/tldr-pages/tldr/pull/15253#discussion_r1932827989
- maybe something else entirely?

I personally like the original options more.

Comment (spageektti) [reply CLIENT-SPECIFICATION.md:43]: I prefer the first proposition, but why do we use `V` and not `L`?

Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: Because -L already in use.

File: CLIENT-SPECIFICATION.md [hunk 6]
```diff
@@ -54,6 +56,17 @@ tldr --version
...
+### Environment variables
+
+The following environment variables MAY be supported by the client to alter its behavior:
+
+Environment variable | Meaning
+---------------------|----------
+TLDR_PLATFORM        | Specify the platform to be used by tldr
+TLDR_LANGUAGE        | Specify which language to prefer
+TLDR_SHORTFORM       | Specify whether to show shortform options
+TLDR_LONGFORM        | Specify whether to show longform options
```
[the remark this answers is not in the document]
Comment (Managor) [reply CLIENT-SPECIFICATION.md:68]: I'll drop TLDR_LANGUAGE and TLDR_PLATFORM from this PR. Lets leave the consideration for those in the next spec version. 
I do like the the last suggestion. I'm going to shorten it to `TLDR_OPTIONS=short/long/both`

File: CLIENT-SPECIFICATION.md (hunk 5 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: A lot of support for the original proposal. I'll ask the question again based on the suggestion sbrl made:
- `-S` for short, `-V` for long options (original proposal)
- `-s` for short, `-e` for long options (alternative idea)(e for expanded)
@acuteenvy @sebastiaanspeck @spageektti @sbrl 
I personally like the alternative idea since it wouldn't involve uppercase characters.

Comment (kbdharun) [reply CLIENT-SPECIFICATION.md:43]: I was going through our official clients and popular community clients, IG we can use `-S` for short form options generally (since `-s` is used for search [node client] and source [python client] options). 

For long options, `-E` would work the best since `-L` and `-l` are used for language and list pages options in clients (also to match case).

So my suggestion would be:

```suggestion
`-S`, `--shortform`| No          | If set, will filter examples to show their shortform option when available
`-E`, `--longform` | No          | If set, will filter examples to show their longform option when available
```

Maybe we can introduce another option to display both options when user overrides this behaviour using the environment variable:

```suggestion
`-S`, `--shortform`| No          | If set, will filter examples to show their shortform option when available
`-E`, `--longform` | No          | If set, will filter examples to show their longform option when available
`-C`, `--complete` | No          | If set in the presence of `TLDR_OPTIONS` environment variable with `short` or `long` variable, will display the default behavior of both the options  
```

`-c` is already used to clear cache across clients, so we can use `-C`.

**Review by kbdharun — APPROVED** [review]
Went through the overall additions (thanks for working on these major changes) and it LGTM after fixing the short options for the new additions.

As informed in the chatroom, I will create a PR later today to document steps in the maintainer guide to create a client specification release.

File: CLIENT-SPECIFICATION.md (hunk 5 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: I think just `-S` and `-E` are enough. The options should override any environment variables. As a sidenote, the default behavior is set to be to only show longform options https://github.com/tldr-pages/tldr/blob/5b75da87f380f7c452fb2c64c4753e7746b0b4bc/CLIENT-SPECIFICATION.md?plain=1#L45

**Review by kbdharun — APPROVED** [review]

File: CLIENT-SPECIFICATION.md [hunk 7]
```diff
@@ -54,6 +56,14 @@ tldr --version
 tldr -l
 ```
 
+### Environment variables
+
+The following environment variables MAY be supported by the client to alter its behavior:
+
+Environment variable | Values          | Meaning
+---------------------|-----------------|----------
+TLDR_OPTIONS         | short/long/both | Specify the option style to display.
```
Comment (acuteenvy) [inline CLIENT-SPECIFICATION.md:65]: How about just saying that the clients may support additional methods of configuration? It's fine to implement `TLDR_OPTIONS` in the Python client, because it's configured using enviroment variables. However I can't help but think that putting env vars in the client spec suggests that this is the preferred way of configuration. There are clients that use a config file for example, and implementing configuration via environment variables in such a client would be quite pointless.

File: CLIENT-SPECIFICATION.md [hunk 8]
```diff
@@ -113,14 +123,26 @@ The structure inside these translation folders is identical to that of the main
...
+Although this specification is about the interface that clients must provide, it is also worth noting that pages are written in standard [CommonMark](https://commonmark.org/), with the exceptions:
+- Non-standard `{{`, `}}`, `{{[` and `]}}` placeholder syntax.
+  - `{{` and `}}` surrounds values in an example that indicates editable values.
+  - `{{[` and `]}}` indicate shortform and longform variants of options which are separated by a single `|`. Shortform on the left, longform on the right.
+
+Things to take into account:
+- Clients MAY highlight the placeholders and MUST remove the surrounding curly braces.
+- Clients MUST remove the angle brackets from option placeholders when only short- or longform is shown and MAY remove them when both are shown depending whether it's clear where the the option placeholder starts and ends (for example when underlining placeholders).
```
Comment (acuteenvy) [inline CLIENT-SPECIFICATION.md:133]: Whether it's clear where the option placeholder starts and ends is subjective. A client could be underlining placeholders depending on user configuration.
We can:
- standardize a single way of displaying both option styles;
- just say that clients may or may not remove them, and may implement a configuration option for whether they'll be removed.

What do you think about this?

File: CLIENT-SPECIFICATION.md [hunk 9]
```diff
@@ -39,8 +39,10 @@ Option             | Required?   | Meaning
 `-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if caching is supported.
 `-l`, `--list`     | No          | Lists all the pages in the current platform to the standard output.
 `-L`, `--language` | No          | Specifies the preferred language for the page returned. Overrides other language detection mechanisms. See the [language section](#language) for more information.
+`-S`, `--shortform`| No          | If set, will filter examples to show their shortform option when available
+`-E`, `--longform` | No          | If set, will filter examples to show their longform option when available
 
-Clients MUST implement both the short and long versions of an option.
+By default clients SHOULD display only the longform option when neither `--shortform` or `--longform` is set by the user. If both are provided, both patterns should be displayed.
```
Comment (acuteenvy) [inline CLIENT-SPECIFICATION.md:45]:
```suggestion
By default clients SHOULD display only the longform option when neither `--shortform` or `--longform` is set by the user. If both are provided, both options should be displayed.
```

File: CLIENT-SPECIFICATION.md
Comment (kbdharun) [inline CLIENT-SPECIFICATION.md:1]: While we are at this can you modify line 38, to include that passing `-p common` should also be supported since we use example like this (across pages):

image

But it won't work in official clients like Node, etc and in community clients like tealdeer.

---

This change was discussed a while ago in the chatroom at https://matrix.to/#/!zXiOpjSkFTvtMpsenJ:gitter.im/$4J7H1ryHsTkjuKuTgZsk19d7EyAk7Fc6TtDBnoFomYw?via=gitter.im&via=matrix.org&via=one.ems.host

File: CLIENT-SPECIFICATION.md (hunk 7 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:65]: Understood. I'll remove that section.

File: CLIENT-SPECIFICATION.md
Comment (Managor) [reply CLIENT-SPECIFICATION.md:1]: I assume changing the "Required" from "Yes" to "No" is what you were after. Matrix links work really poorly so I'm unable to check what the discussion was about.

File: CLIENT-SPECIFICATION.md (hunk 8 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:125]: I do not think it's subjective. Again, see here for examples https://github.com/tldr-pages/tldr-python-client/pull/259#issuecomment-2619942873. It is not fully clear where the longform options end unless you're using a client like tealdeer which underlines each placeholder. 
image
image

Comment (acuteenvy) [reply CLIENT-SPECIFICATION.md:125]: Underlining placeholders can be configurable. If a client allows it, should the brackets be removed depending on the option? That would be a weird unexpected side effect in my opinion. It'd be better to have a dedicated configuration option.

Comment (Managor) [reply CLIENT-SPECIFICATION.md:125]: I just think that it should be clear where the option placeholder starts and ends. I don't understand why it would need a dedicated configuration option.

File: CLIENT-SPECIFICATION.md
Comment (acuteenvy) [reply CLIENT-SPECIFICATION.md:1]: > I assume changing the "Required" from "Yes" to "No" is what you were after. Matrix links work really poorly so I'm unable to check what the discussion was about.

This was about supporting `common` as a possible argument for `-p`/`--platform`.

File: CLIENT-SPECIFICATION.md (hunk 8 above)
Comment (acuteenvy) [reply CLIENT-SPECIFICATION.md:125]: > I don't understand why it would need a dedicated configuration option.

If a client currently has a configuration option to underline placeholders, then by your specification, that same option now also sets the rules for displaying square brackets. So why not make it a separate option? Someone might want a different combination (i.e. square brackets with underlining or no square brackets and no underlining).

File: CLIENT-SPECIFICATION.md [hunk 10]
```diff
@@ -30,17 +30,19 @@ This section describes the standardised command-line interface (CLI) for clients
...
-`-p`, `--platform` | Yes         | Specifies the platform to be used to perform the action (either listing or searching) as an argument. If this option is specified, the selected platform MUST be checked first instead of the current platform as described below.
+`-p`, `--platform` | No         | Specifies the platform to be used to perform the action (either listing or searching) as an argument. If this option is specified, the selected platform MUST be checked first instead of the current platform as described below.
 `-u`, `--update`   | Conditional | Updates the offline cache of pages. MUST be implemented if caching is supported.
 `-l`, `--list`     | No          | Lists all the pages in the current platform to the standard output.
 `-L`, `--language` | No          | Specifies the preferred language for the page returned. Overrides other language detection mechanisms. See the [language section](#language) for more information.
+`-S`, `--shortform`| No          | If set, will filter examples to show their shortform option when available
+`-E`, `--longform` | No          | If set, will filter examples to show their longform option when available
```
Comment (acuteenvy) [inline CLIENT-SPECIFICATION.md:43]: I would like to propose changing the long options too. With the current ones (i.e `--shortform` and `--longform`) it's impossible to know what exactly the options are doing (short/long form _**of what?**_) without checking the documentation.

What about `--short-options` and `--long-options`?

File: CLIENT-SPECIFICATION.md (hunk 8 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:125]: If the uniformity is a problem, we can just make it so that the brackets are always visible when both short and long options are shown.

Comment (acuteenvy) [reply CLIENT-SPECIFICATION.md:125]: > If the uniformity is a problem, we can just make it so that the brackets are always visible when both short and long options are shown.

That was my first suggestion:
> - standardize a single way of displaying both option styles;

File: CLIENT-SPECIFICATION.md
Comment (Managor) [reply CLIENT-SPECIFICATION.md:1]: Is that not documented in here?
https://github.com/tldr-pages/tldr/blob/ca61c713ed7b1b21069fec2ba527e0868876ef72/CLIENT-SPECIFICATION.md?plain=1#L21
It's not up to the specifications to document clients that don't follow the spec.

File: CLIENT-SPECIFICATION.md (hunk 8 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md]: Alright, my bad. The bikeshedding is getting to me.

File: CLIENT-SPECIFICATION.md
Comment (Managor) [reply CLIENT-SPECIFICATION.md:1]: @kbdharun Can you instead create a separate PR after this instead. I don't know what you mean and I just want this PR to be over. It'll be included in the same version as we're not making the release before it's finished.

File: CLIENT-SPECIFICATION.md (hunk 8 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md]: >If a client currently has a configuration option to underline placeholders, then by your specification, that same option now also sets the rules for displaying square brackets

I didn't sees this one. No, the underline is not a configuration option. It's the default rendering style in tealdeer.

**Review by acuteenvy — APPROVED** [review]

**Comment by Managor** [comment]
@sbrl @spageektti May I get your approval for the changes?

**Review by kbdharun — APPROVED** [review]

**Review by gutjuri — APPROVED** [review]

**Comment by sbrl** [comment]
Oops! Just thought of 1 more thing we need to do before merging, @Managor. Could you update the changelog at the bottom of the file please?

**Comment by Managor** [comment]
At least there no decision to be made with the changelog. It's just do and done addition.

**Review by sbrl — APPROVED** [review]
Thanks for your hard work on this, @Managor! Are there any client maintainers we can ping for a final opinion before we merge this?

**Comment by Managor** [comment]
I assume @niklasmohrin would be a good contact point for tealdeer.

File: CLIENT-SPECIFICATION.md [hunk 11]
```diff
@@ -30,17 +30,19 @@ This section describes the standardised command-line interface (CLI) for clients
...
+Option                 | Required?   | Meaning
+-----------------------|-------------|----------
+`-v`, `--version`      | Yes         | Shows the current version of the client, and the version of this specification that it implements.
+`-p`, `--platform`     | Yes         | Specifies the platform to be used to perform the action (either listing or searching) as an argument. If this option is specified, the selected platform MUST be checked first instead of the current platform as described below.
+`-u`, `--update`       | Conditional | Updates the offline cache of pages. MUST be implemented if caching is supported.
+`-l`, `--list`         | No          | Lists all the pages in the current platform to the standard output.
+`-L`, `--language`     | No          | Specifies the preferred language for the page returned. Overrides other language detection mechanisms. See the [language section](#language) for more information.
+`-S`, `--short-options`| No          | If set, will filter examples to show their shortform option when available
+`-E`, `--long-options` | No          | If set, will filter examples to show their longform option when available
```
Comment (niklasmohrin) [inline CLIENT-SPECIFICATION.md:43]: I think `tldr -E tar` looks rather strange, my feeling would be that users would usually have an alias for their preferred display style like `alias tldr=tldr -E` anyways, so having that as `alias tldr=tldr --long-options` doesn't really hurt.

In the context of tealdeer, I would expect users to set their preference in the tealdeer config file and be mostly done with it.

Wouldn't it suffice to only require the long versions of these flags for now, and see if users ask for short versions?

File: CLIENT-SPECIFICATION.md [hunk 12]
```diff
@@ -30,17 +30,19 @@ This section describes the standardised command-line interface (CLI) for clients
...
 
-Clients MUST implement both the short and long versions of an option.
+By default clients SHOULD display only the longform option when neither `--short-options` or `--long-options` is set by the user. If both are provided, both options should be displayed.
```
Comment (niklasmohrin) [inline CLIENT-SPECIFICATION.md:45]: How do you feel about something like this?
```suggestion
By default clients SHOULD display only the longform option when neither `--short-options` or `--long-options` is set by the user. If both are provided, both options should be displayed (see the Page Structure / Examples section for the output format).
```
Otherwise, this sentence leaves me wondering what I should print when both options are passed. Alternatively, an example could also be provided right here

**Review by niklasmohrin — COMMENTED** [review]
Cool idea!

I don't really like the `-S` and `-E` options, and I would prefer sticking to the long options - at least until we get some feedback from users using the long versions

File: CLIENT-SPECIFICATION.md (hunk 11 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: In the text below I specified that clients should default to showing longform options, but can show shortform by default if the client wants to advertise itself as power users first or they can choose to default to showing both. This makes this change backwards compatible as a client that doesn't update is still within the spec.
https://github.com/tldr-pages/tldr/blob/c762eb77ea4093d2bbf35952fe221866a055d122/CLIENT-SPECIFICATION.md?plain=1#L45
> Wouldn't it suffice to only require the long versions of these flags for now, and see if users ask for short versions?

As the table above shows, neither of these options is required. You can implement these at your own pace.

Comment (niklasmohrin) [reply CLIENT-SPECIFICATION.md:43]: I didn't mean to criticize the default behavior, only that I don't see too much value in requiring the short forms `-S` and `-E`. In particular, `-E` seems somewhat random and not worth it to me, as I would expect that users usually configure this once and not on every run. So maybe, in my example, let's say users have `alias tldr=tldr -SE`. My point is that in the alias, it doesn't hurt to write `alias tldr=tldr --short-options --long-options`, and on the command line, I don't know if people would use `tldr -SE` often enough to make it worth the effort.

Still, I see a lot of value in having `--short-options` and `--long-options` in general and think that the configuration through the long form options is modeled nicely.

The reason I suggest to only include the long options in the spec is that we also have the following:

> If an option is implemented both the short and long versions MUST be available

Thus, having only `--long-options` would not be compliant.

Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: I see. That requirement was there before this patch, it just got moved to the start to prevent confusion. Changing the requirement would be a separate issue from this in my opinion. If you could make an issue or a pull request, we can have a discussion and then bundle it with the 2.3 release.

**Comment by Managor** [comment]
Unless anything major happens, I will merge this PR on Sunday. If you know any maintainers who might want to chime in, please reach out to them before that.

After this, we'll have pull requests and discussions about dropping the requirement to have both shortform and longform options available https://github.com/tldr-pages/tldr/pull/15253#discussion_r1965450746 and including `common` as a platform https://github.com/tldr-pages/tldr/pull/15253#discussion_r1957223790

File: CLIENT-SPECIFICATION.md (hunk 11 above)
Comment (niklasmohrin) [reply CLIENT-SPECIFICATION.md:43]: We are talking about different things again. I am fine with "If an option is implemented both the short and long versions MUST be available". What I am suggesting is the following:
```suggestion
`--short-options`| No          | If set, will filter examples to show their shortform option when available
`--long-options` | No          | If set, will filter examples to show their longform option when available
```
My rationale is that we can still introduce `-S` and `-E` in a later release, if we feel that it makes sense, but if we add it to the spec now, we will not be able to take it back later. I think that they are unneeded and we might regret introducing them in the next release already.

File: CLIENT-SPECIFICATION.md
Comment (niklasmohrin) [reply CLIENT-SPECIFICATION.md:1]: It will work in tealdeer starting in the next release :) https://github.com/tealdeer-rs/tealdeer/pull/401

File: CLIENT-SPECIFICATION.md (hunk 11 above)
Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: The PR completing doesn't mean we make it official yet. The release comes later. It would be better to have this discussion in a separate PR.

Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: I changed the version number to `Unreleased` so there's less chances of someone saving the in-between version and thinks it's the official release.

Comment (niklasmohrin) [reply CLIENT-SPECIFICATION.md:43]: I am again not sure if we are talking about the same thing. From your responses, it feels like you want to make a separate discussion about the "all variants of a tldr flag should be implemented" rule. However, I don't mean to change the general rule. I am only suggesting that we add just the long form variant for the new `--short-options` and `--long-options` flags. Thus, I don't see why we should merge this PR and then make a separate one about removing some of the additions again.

For context, I assume that the intention behind "Clients MUST implement both the short and long versions of an option" is that all listed variants in the table should be implemented. Thus, it would not impose that all flags actually must _have_ multiple variants. Whether or not this interpretation of the spec is correct is of course up for debate, but I think there is no precedence for either interpretation.

Anyways, I don't want to pretend that I am in the position to block this PR with my concerns. Still, I feel that my actual point has so far only been dismissed in what feels like a simple misunderstanding. It would be nice if you could reconsider my previous comments with that in mind

Comment (Managor) [reply CLIENT-SPECIFICATION.md:43]: I want it in a separate PR/issue, because I want this long running PR to be over and because I want the discussion to not be drowned in this discussion. In my opinion, this change would be a major break from form to not have a shortform option.

Comment (niklasmohrin) [reply CLIENT-SPECIFICATION.md:43]: Okay then

Comment (niklasmohrin) [reply CLIENT-SPECIFICATION.md:43]: See #15772
